### PR TITLE
Remove Guardian Management deps

### PR DESCRIPTION
**Pull request by nicl**
These dependencies (https://github.com/guardian/guardian-management) were last updated 4 years ago, very little used outsideo of RR (it looks like RR and Flexible Content are the two main users from a quick Github search) and have proven very difficult to update to Scala 2.13+.

As such, we are removing them for security and maintenance reasons.

Switch-related code has been partially inlined from the library as part of this to ease the transition.

Metrics, which were done using Guardian Management and served to a JSON endpoint for consumption by Ganglia (which no longer exists) have been removed completely for now.

A fuller switch solution would be useful here, but is beyond the scope of this immediate change.

**UPDATE: tested on CODE successfully.**

### CODE REVIEW

File: magenta-lib/src/main/scala/magenta/deployment_type/SelfDeploy.scala [hunk 1]
```diff
@@ -53,26 +53,16 @@ object SelfDeploy extends DeploymentType {
...
-      ChangeSwitch(
-        host,
-        managementProtocol(pkg, target, reporter),
-        managementPort(pkg, target, reporter),
-        switchboardPath(pkg, target, reporter),
-        "shutdown-when-inactive",
-        desiredState=true
-      )
-    }
+    hosts.map(ShutdownTask.apply)
```
Comment (nicl) [inline magenta-lib/src/main/scala/magenta/deployment_type/SelfDeploy.scala:65]: Guardian Management automatically presented endpoints to toggle switches on a non-standard port. This has been replaced with a custom endpoint (on the standard 9000) to do this.

File: magenta-lib/src/main/scala/magenta/logging.scala [hunk 2]
```diff
@@ -190,4 +187,8 @@ case object Message extends Enum[Message] with PlayJsonEnum[Message] {
   case class Warning(text: String) extends Message
 
   val values = findValues
+}
+
+trait Loggable {
```
Comment (nicl) [inline magenta-lib/src/main/scala/magenta/logging.scala:192]: A basic `Loggable` trait as a drop-in replacement for the same-named trait in Guardian Management.

File: magenta-lib/src/main/scala/magenta/switches.scala [hunk 3]
```diff
@@ -0,0 +1,65 @@
+// Modified version of https://github.com/guardian/guardian-management/blob/c7d99120927958757b35bea68931c4bd8690d64a/management/src/main/scala/com/gu/management/switchables.scala
```
Comment (nicl) [inline magenta-lib/src/main/scala/magenta/switches.scala:1]: Copied from Guardian Management as rewriting switch logic would have involved wider changes. Not all of the switch logic from Guardian Management has been copied though, only the basic switch models.

File: riff-raff/app/conf/context.scala [hunk 4]
```diff
@@ -269,98 +255,3 @@ class Config(configuration: TypesafeConfig, startTime: DateTime) extends Logging
 
   override def toString: String = configuration.toString
 }
-
-class Management(config: Config, shutdownWhenInactive: ShutdownWhenInactive, deployments: Deployments, datastore: DataStore) {
```
Comment (nicl) [inline riff-raff/app/conf/context.scala:273]: We collected a load of metrics using Guardian Management that look interesting but aren't actually in use from what I can tell. Open Telemetry/something else is probably the future here.

File: riff-raff/app/deployment/Deployments.scala [hunk 5]
```diff
@@ -36,8 +35,6 @@ class Deployments(deploymentEngine: DeploymentEngine,
       requestedParams.build.projectName, requestedParams.stage.name, requestSource)
     if (restrictionsPreventingDeploy.nonEmpty) {
       Left(Error(s"Unable to queue deploy as restrictions are currently in place: ${restrictionsPreventingDeploy.map(r => s"${r.fullName}: ${r.note}").mkString("; ")}"))
-    } else if (enableQueueingSwitch.isSwitchedOff) {
```
Comment (nicl) [inline riff-raff/app/deployment/Deployments.scala:39]: (I have not recreated the enableQueueing switch as was not convinced of the value, but we can re-add if it's more useful than I realise.)

**Review by akash1810 — COMMENTED** [review]
These changes look sensible, great work. I'm a little surprised that we've been collecting all these metrics w/out displaying them anywhere, which leads me to think they might have another, subtle, use other than fancy graphs. I can't quite think of what though. Let's run this on CODE to test.

**Comment by akash1810** [comment]
> UPDATE: tested on CODE successfully.

How exhaustively did we test? Do we need to do each deployment type, for example? (Probably not)

File: magenta-lib/src/main/scala/magenta/tasks/tasks.scala [hunk 6]
```diff
@@ -209,40 +208,35 @@ case class SayHello(host: Host)(implicit val keyRing: KeyRing) extends Task {
...
-  val switchboardUrl = s"$protocol://${host.name}:$port$path"
+case class ShutdownTask(host: Host)(implicit val keyRing: KeyRing)  extends Task {
 
-  // execute this task (should throw on failure)
-  override def execute(resources: DeploymentResources, stopFlag: => Boolean): Unit = {
-    resources.reporter.verbose(s"Changing $switchName to $desiredStateName using $switchboardUrl")
+  val description = "Shutdown Riffraff target(s) as part of self-deploy."
 
+  override def execute(resources: DeploymentResources, stopFlag: => Boolean): Unit = {
+    val shutdownUrl = s"https:://${host.name}:9000/shutdown"
```
[the remark this answers is not in the document]
Comment (nicl) [reply magenta-lib/src/main/scala/magenta/tasks/tasks.scala:216]: @akash1810 this actually gets used when deploying PROD Riffraff from CODE and vice versa, so is potentially across the internet. We could probably figure out if it was targeting itself, but I'm not sure it's worth it really.

**Comment by nicl** [comment]
@akash1810 it definitely wasn't an exhaustive test. I basically clicked through the main nav links to check they worked and also did a self-deploy.

**Comment by nicl** [comment]
@akash1810 I am definitely suspicious that the metrics were used elsewhere too! I'll have another look to see if I can find anything else.

File: riff-raff/app/deployment/Deployments.scala [hunk 7]
```diff
@@ -143,6 +125,22 @@ class Deployments(deploymentEngine: DeploymentEngine,
 
   lazy val enableQueueingSwitch = new DefaultSwitch("enable-deploy-queuing", "Enable riff-raff to queue deploys.  Turning this off will prevent anyone queueing a new deploy, although running deploys will continue.", true)
```
Comment (jacobwinch) [inline riff-raff/app/deployment/Deployments.scala:126]: Based on [this comment](https://github.com/guardian/riff-raff/pull/663/files#r769648481) I think this switch can be removed now?

File: riff-raff/app/deployment/Deployments.scala (hunk 5 above)
Comment (jacobwinch) [reply riff-raff/app/deployment/Deployments.scala:39]: I don't remember this being used in my time on DevX so I think it's OK to remove it.

File: riff-raff/app/deployment/Deployments.scala [hunk 8]
```diff
@@ -91,27 +89,11 @@ class Deployments(deploymentEngine: DeploymentEngine,
...
-      } catch {
-        case NonFatal(t) =>
-          log.error(s"Exception thrown whilst processing update with $wrapper", t)
-      }
-    }
-  def init() {}
-  def shutdown() { messagesSubscription.unsubscribe() }
-
-
   val deploysEnabled = new Switch(startAsOn = true)
```
Comment (jacobwinch) [inline riff-raff/app/deployment/Deployments.scala:92]: I noticed that this is using `akka.util.Switch`, which seems to provide pretty similar functionality to some of the code that is being ported across from Guardian Management - I wonder if we could simplify/make this consistent in the future

File: magenta-lib/src/main/scala/magenta/deployment_type/SelfDeploy.scala (hunk 1 above)
Comment (jacobwinch) [reply magenta-lib/src/main/scala/magenta/deployment_type/SelfDeploy.scala:55]: I guess this allows us to remove [this ingress rule](https://github.com/guardian/deploy-tools-platform/blob/ea015f04e297291ad7b2066095459c273b8e447e/cloudformation/riffraff/riffraff.template.yaml#L141-L146)?

File: magenta-lib/src/main/scala/magenta/tasks/tasks.scala [hunk 9]
```diff
@@ -209,40 +208,39 @@ case class SayHello(host: Host)(implicit val keyRing: KeyRing) extends Task {
...
-case class ChangeSwitch(host: Host, protocol:String, port: Int, path: String, switchName: String, desiredState: Boolean)(implicit val keyRing: KeyRing) extends Task {
-  val desiredStateName: String = if (desiredState) "ON" else "OFF"
-  val switchboardUrl = s"$protocol://${host.name}:$port$path"
+case class ShutdownTask(host: Host)(implicit val keyRing: KeyRing)  extends Task {
 
-  // execute this task (should throw on failure)
-  override def execute(resources: DeploymentResources, stopFlag: => Boolean): Unit = {
-    resources.reporter.verbose(s"Changing $switchName to $desiredStateName using $switchboardUrl")
+  val description = "Shutdown Riffraff target(s) as part of self-deploy."
+  val shutdownUrl = s"http://${host.name}:9000/shutdown"
```
Comment (jacobwinch) [inline magenta-lib/src/main/scala/magenta/tasks/tasks.scala:214]: Should this be `/requestShutdown`?

File: magenta-lib/src/main/scala/magenta/tasks/tasks.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (nicl) [reply magenta-lib/src/main/scala/magenta/tasks/tasks.scala]: Updating this, it turns out that there is actually a manual deployment restriction on cross-stage deployments of Riffraff as, while the code itself supports this behaviour, the security groups do not.

File: riff-raff/app/deployment/Deployments.scala (hunk 8 above)
Comment (nicl) [reply riff-raff/app/deployment/Deployments.scala:92]: Yeah, I think so. It was surprising to me that both were used. This PR is already quite large(!) but this is a good candidate for a follow-up.

**Review by jacobwinch — APPROVED** [review]
👍

**Pull request merged**